src/repository/opamRepositoryBackend.mli

```diff
@@ -56,6 +56,13 @@ module type S = sig
     repository_name -> ?cache_dir:dirname -> dirname -> url ->
     update OpamProcess.job
 
+  (** [repo_update_complete dirname url] finalizes the update of the repository
+      after verification of the patch returned from [pull_repo_update] with
+      [Update_patch file] is applied. Version control systems, e.g. Mercurial,
+      that track the state of the working directory automatically use this to
+      update internal caches. *)
+  val repo_update_complete: dirname -> url -> unit OpamProcess.job
```

Comment:
I tried mightily to solve this problem without extending the interface, and I share your dissatisfaction with it. I agree that `VCS.reset` has an unclear name and that another function, which for Git performs at most `git reset --soft` and which minimally is a null function, would be a less intrusive change.

Alas, Mercurial does not expose [its internal directory](https://www.mercurial-scm.org/wiki/DirState) state like Git does with its "index" concept. There is no equivalent of the `git reset --soft` command, and this is quite a deliberate choice by the Mercurial team. So when OPAM applies the patch after validating it, the Mercurial directory state remains unchanged, leaving it continuing to reference the revision prior to the update. This is why the next OPAM update fails— the directory state shows uncommitted changes (those applied by the previous update's patch).

Here's what I propose to do:

- Change the name of the existing `VCS.reset` function to `VCS.revert` (which incidentally more closely matches the generally more carefully considered Mercurial terminology, i.e. it reverts the working directory to its unchanged state).

- Add a new `VCS.patch_applied` function that is a no-operation for all the other VCS implementations (with a comment in `opamGit.ml` referencing this PR# that says `git reset --soft` may be appropriate there). For Mercurial, this new function would invoke `hg update --clean --rev`

- Call `VCS.patch_applied` instead of `VCS.reset` accordingly.

On the related topic of cryptographic signing of repository updates, I wonder if the right way forward for that is to rely on the native support for that in supported version control systems, e.g. Git signed commits, the Mercurial GPG extension, et cetera.

Reply:
> This is why the next OPAM update fails— the directory state shows uncommitted changes (those applied by the previous update's patch).

I think we are on the right track here, but just to be sure everything has been considered: could doing the `hg update` at the beginning of `VCS.fetch` be a solution ?

Otherwise, that sounds perfect. Maybe be even more explicit with a name like `VCS.reset_tree` ?

For the signatures, we have considered git signed commits and the like, but they aren't a perfect fit for end-to-end encryption of the package definitions we wanted (as opposed to an authority signing the repository as a whole). If you are interested, see the work-in progress here: https://github.com/hannesm/conex (the README has pointers to the related papers)

Reply:
Alas, `hg update` always has the potential to update the working directory state— that's why it's called the "update" command. It is an invariant of the Mercurial repository format that its `DirState` is an accurate record of what is the pending to commit for the next revision, and the method OPAM uses to fetch new revisions, generate a patch, validate it then apply it, pretty much breaks Mercurial's invariant.

An alternative that might work: have `VCS.fetch` first perform `hg pull` then generate the patch, then `hg update --rev` followed by `hg revert --no-backup --rev` back to the revision prior to the fetch. In this way, Mercurial will overwrite the working directory twice: firstly) to update to the latest revision, and secondly) to revert the files back to the earlier revision without modifying the `DirState`. At the point, applying the generated patch will have the effect of reverting the effect of the `hg revert` command. Is that better? I dunno. You tell me.

Reply:
Well, crap. It turns out there are unpleasant cases where `hg revert -Ca -r ${id}` does not work, because it fails when you try to revert to a revision before the introduction of a subrepository.

Reply:
I'm going to push another revision that addresses some of the minor nits flagged in review comments above, and which also makes the name changes discussed above. Tomorrow or the next day, I'll try another approach that explicitly applies the generated patch with `patch -R` before returning it from `VCS.fetch` to be first validated, then reapplied in the forward direction, thereby restoring the working directory to the clean update state.

I don't think I like how complicated this seems necessary to be, but I'm not seeing any simpler way to get this done right.

Reply:
The apply-then-revert approach does indeed sound a little brittle, considering signing and security; and calling another sync function after applying the patch is a pretty fine solution considering the concerns above.
Thanks again!